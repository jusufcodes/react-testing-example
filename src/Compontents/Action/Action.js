import React, {useState} from "react";
import ShareIcon from "../Icon/Share.svg";
import FavoriteIcon from "../Icon/Favorite.svg";
import classNames from "classnames";

export const Action = (props) => {
  const [clickCount, setClickCount] = useState(0);

  const handleClick = (event) => {
    setClickCount(clickCount + 1);

    if (props.onClick) {
      props.onClick(event);
    }
  };

  return props.children ? (
    <button onClick={handleClick}>
      {props.children} ({clickCount})
    </button>
  ) : null;
};

export async function likeAndShare(postId) {
  const apiUrl = postId ? "posts" : "error";

  const response = await fetch(`https://jsonplaceholder.typicode.com/${apiUrl}`, {
    method: 'POST',
    body: JSON.stringify({
      id: postId
    }),
    headers: {
      "Content-type": "application/json; charset=UTF-8"
    }
  });

  const json = await response.json();

  return json.id;
}

export const FavoriteAction = (props) => {
  const [liked, setLike] = useState(false);

  async function handleClick() {
    const currentLiked = liked;

    setLike(!currentLiked);

    if (!props.postId) {
      return
    }

    const id = await likeAndShare(props.postId);

    if (typeof id === "number") {
      setLike(!currentLiked);
    } else {
      alert("error");
    }
  }

  return props.postId ? (
    <Action onClick={handleClick}>
      <img src={FavoriteIcon} alt="" className={classNames("favoriteIcon", {
        active: liked
      })}/>
    </Action>
  ) : null;
};

export const ShareAction = (props) => {
  const [shared, setShared] = useState(false);

  async function handleClick() {
    const currentShared = shared;

    setShared(!currentShared);

    if (!props.postId) {
      return
    }

    const id = await likeAndShare(props.postId);

    if (typeof id === "number") {
      setShared(!currentShared);
    } else {
      alert("error");
    }
  }

  return props.postId ? (
    <Action onClick={handleClick}>
      <img src={ShareIcon} alt="" className={classNames("shareIcon", {
        active: shared
      })}/>
    </Action>
  ) : null;
};