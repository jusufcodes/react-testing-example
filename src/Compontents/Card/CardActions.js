import React from "react";

export const CardActions = (props) => {
  return (props.left || props.right) ? (
    <div className={"cardActions"}>
      {props.left && (
        <div className={"cardActionsLeft"}>
          {props.left}
        </div>
      )}
      {props.right && (
        <div className={"cardActionsRight"}>
          {props.right}
        </div>
      )}
    </div>
  ) : null;
};