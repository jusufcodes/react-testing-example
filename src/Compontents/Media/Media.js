import React from "react";

export const Media = (props) => {
  return props.src ? <img src={props.src} alt={props.alt ? props.alt : ""} /> : null;
};
