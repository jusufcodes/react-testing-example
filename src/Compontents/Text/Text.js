import React from "react";

export const Text = (props) => {
  return props.children ? <p>{props.children}</p> : null;
};