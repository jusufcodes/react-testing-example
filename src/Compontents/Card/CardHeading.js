import React from "react";
import {Media} from "../Media/Media";
import {Heading} from "../Heading/Heading";

export const CardHeading = (props) => {
  return (props.avatar || props.heading || props.subHeading) ? (
    <div className={"cardHeading"}>
      {props.avatar && <Media src={props.avatar} />}
      {props.heading && <Heading level={2}>{props.heading}</Heading>}
      {props.subHeading && <Heading level={3}>{props.subHeading}</Heading>}
    </div>
  ) : null;
};