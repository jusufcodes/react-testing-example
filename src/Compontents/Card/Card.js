import React from "react";
import {CardHeading} from "./CardHeading";
import {Media} from "../Media/Media";
import {Text} from "../Text/Text";
import {CardActions} from "./CardActions";

export const Card = (props) => {
  return (props.heading || props.media || props.text || props.actions) ? (
    <div className={"card"}>
      {props.heading && (
        <CardHeading
          avatar={props.heading.avatar}
          heading={props.heading.heading}
          subHeading={props.heading.subHeading}
        />
      )}

      {props.media && (
        <Media src={props.media} />
      )}

      {props.text && (
        <Text>{props.text}</Text>
      )}

      {props.actions && (
        <CardActions
          {...props.actions}
        />
      )}
    </div>
  ) : null;
};