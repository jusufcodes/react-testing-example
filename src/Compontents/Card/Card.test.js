import React from 'react';
import {shallow} from 'enzyme';
import renderer from "react-test-renderer";
import {Card} from "./Card";
import {CardHeading} from "./CardHeading";
import {Media} from "../Media/Media";
import {CardActions} from "./CardActions";
import {Action, FavoriteAction, ShareAction} from "../Action/Action";
import {Text} from "../Text/Text";

const demoHeading = {
  avatar: "https://via.placeholder.com/60x60",
  heading: "Card Example",
  subHeading: "Demonstration of TDD"
};

const demoMedia = "https://via.placeholder.com/300x200";
const demoText = "Lorem ipsum dolor sit amet";
const demoAction = {
  left: (
    <>
      <Action>Action 1</Action>
      <Action>Action 2</Action>
    </>
  ),
  right: (
    <>
      <FavoriteAction/>
      <ShareAction/>
    </>
  )
};

describe("Card component", () => {

  test("Test without props", () => {
    const card = <Card/>;

    expect(shallow(card).isEmptyRender()).toEqual(true);
  });

  test("Test with heading", () => {
    const card = <Card heading={demoHeading}/>;
    const shallowCard = shallow(card);

    expect(shallowCard.isEmptyRender()).toEqual(false);
    expect(shallowCard.children().length).toEqual(1);
    expect(shallowCard.find(CardHeading).length).toEqual(1);

    const snapshot = renderer.create(card).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  test("Test with media", () => {
    const card = <Card media={demoMedia}/>;
    const shallowCard = shallow(card);

    expect(shallowCard.isEmptyRender()).toEqual(false);
    expect(shallowCard.children().length).toEqual(1);
    expect(shallowCard.find(Media).length).toEqual(1);

    const snapshot = renderer.create(card).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  test("Test with Text", () => {
    const card = <Card text={demoText}/>;
    const shallowCard = shallow(card);

    expect(shallowCard.isEmptyRender()).toEqual(false);
    expect(shallowCard.children().length).toEqual(1);
    expect(shallowCard.find(Text).length).toEqual(1);

    const snapshot = renderer.create(card).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  test("Test with Actions", () => {
    const card = <Card actions={demoAction}/>;
    const shallowCard = shallow(card);

    expect(shallowCard.isEmptyRender()).toEqual(false);
    expect(shallowCard.children().length).toEqual(1);
    expect(shallowCard.find(CardActions).length).toEqual(1);

    const snapshot = renderer.create(card).toJSON();
    expect(snapshot).toMatchSnapshot();
  });

});
