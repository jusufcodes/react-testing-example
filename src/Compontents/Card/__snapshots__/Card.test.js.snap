// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Test with Actions 1`] = `
<div
  className="card"
>
  <div
    className="cardActions"
  >
    <div
      className="cardActionsLeft"
    >
      <button
        onClick={[Function]}
      >
        Action 1
         (
        0
        )
      </button>
      <button
        onClick={[Function]}
      >
        Action 2
         (
        0
        )
      </button>
    </div>
    <div
      className="cardActionsRight"
    />
  </div>
</div>
`;

exports[`Card component Test with Text 1`] = `
<div
  className="card"
>
  <p>
    Lorem ipsum dolor sit amet
  </p>
</div>
`;

exports[`Card component Test with heading 1`] = `
<div
  className="card"
>
  <div
    className="cardHeading"
  >
    <img
      alt=""
      src="https://via.placeholder.com/60x60"
    />
    <h2>
      Card Example
    </h2>
    <h3>
      Demonstration of TDD
    </h3>
  </div>
</div>
`;

exports[`Card component Test with media 1`] = `
<div
  className="card"
>
  <img
    alt=""
    src="https://via.placeholder.com/300x200"
  />
</div>
`;
